pkg/api/pin_chunks.go

```diff
@@ -312,10 +318,34 @@ func (s *server) updatePinCount(ctx context.Context, reference swarm.Address, de
 
 func (s *server) pinChunkAddressFn(ctx context.Context, reference swarm.Address) func(address swarm.Address) (stop bool) {
 	return func(address swarm.Address) (stop bool) {
-		err := s.Storer.Set(ctx, storage.ModeSetPin, address)
+		// NOTE: stop pinning on first error
+
+		has, err := s.Storer.Has(ctx, address)
+		if err != nil {
+			s.Logger.Debugf("pin traversal: localstore has: for reference %s, address %s: %w", reference, address, err)
+			return true
+		}
+
+		if !has {
+			// chunk not found locally, try to get from netstore
+			ch, err := s.Storer.Get(ctx, storage.ModeGetRequest, address)
+			if err != nil {
+				s.Logger.Debugf("pin traversal: storer get: for reference %s, address %s: %w", reference, address, err)
+				return true
+			}
+
+			_, err = s.Storer.Put(ctx, storage.ModePutRequestPin, ch)
+			if err != nil {
+				s.Logger.Debugf("pin traversal: storer put pin: for reference %s, address %s: %w", reference, address, err)
+				return true
+			}
+
+			return false
+		}
+
+		err = s.Storer.Set(ctx, storage.ModeSetPin, address)
```

Comment:
just an observation - you might have a data race between `L324` and here, and the chunk might be already garbage collected before it can be pinned. I'm trying to think if we need a separate `ModeGet` that also pins a chunk atomically.

Reply:
Yes, that is correct. This was existing way of doing things, and I figured to leave it as it was.

Reason for adding `storage.ModePutRequestPin` mode was to have in in single operation for fetch and pin.
This relates to comment I made here already. We would have to completely read the chunk in order to have it pinned "atomically" (under same lock).

Even though recognize we should have something, I do not completely agree with suggested solution. I do not think that `ModeGet` operation should perform writes in the database. Read operations should only be used to read.

I would rather change the `ModeSetPin` operation to check if the chunk actually exists before adding the pin or return error otherwise, however in that case we also access different indexes.

Reply:
so the only suggestion that comes into mind would to (unfortunately) use `Put` here with a `ModePutUploadPin`, but have `Put` implemented in a way that if the chunk already exists in the database, then it would just increment the pin counter (which is already the case).

Reply:
Yeah, every time I look at `Mode` enums, I think maybe I can bitmask them, and then remember otherwise.

Not sure if you checked updated code (I updated description about the change), it read as:

> Modified storage.ModeSetPin operation, to check if the chunk exists locally before performing pin operation, under same lock (changed Mock storage also). This caused some changes on how the chunk is being pinned.

So basically solution I did here is modify existing operation. My thinking was that it is acceptable to change `storage.ModeSetPin` to expect the chunk to exists locally. If this is not acceptable we could discuss other solutions.